File: components/battery.py
import subprocess
from PyQt6.QtWidgets import QLabel, QWidget
from PyQt6.QtGui import QFont
from PyQt6.QtCore import Qt, QTimer


class BatteryManager:
    def __init__(self, parent):
        self.parent = parent
        self.labels = []
        self.battery_device = self.find_battery_device()
        self.timer = QTimer()
        self.timer.timeout.connect(self.update_info)
        self.init_battery()

    def find_battery_device(self):
        try:
            output = subprocess.check_output(["upower", "-e"]).decode()
            for line in output.strip().splitlines():
                if "battery_" in line:
                    return line.strip()
        except Exception:
            return None

    def parse_battery_info(self, device):
        try:
            output = subprocess.check_output(["upower", "-i", device]).decode()
        except Exception:
            return {}

        info = {}
        wanted_keys = [
            "state", "warning-level", "energy-full", "energy-full-design",
            "energy-rate", "voltage", "charge-cycles", "percentage",
            "capacity", "technology"
        ]
        for line in output.splitlines():
            if ":" not in line:
                continue
            key, value = line.strip().split(":", 1)
            key = key.strip()
            value = value.strip()
            if key in wanted_keys:
                info[key] = value
        return info

    def init_battery(self):
        self.battery_box = QWidget(self.parent)
        self.battery_box.setGeometry(500, 100, 600, 500)
        self.battery_box.setStyleSheet("""
            background-color: rgba(255, 255, 255, 0.05);
            border: 2px solid rgba(255, 255, 255, 0.15);
            border-radius: 30px;
        """)
        self.battery_box.setVisible(False)

        # Title Label
        title = QLabel("BATTERY", self.battery_box)
        title.setFont(QFont("ElysiaOSNew", 20, QFont.Weight.Bold))
        title.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
        title.setAlignment(Qt.AlignmentFlag.AlignCenter)
        title.setGeometry(200, 20, 200, 40)
        title.setVisible(True)

        if not self.battery_device:
            label = QLabel("No battery device found.", self.battery_box)
            label.setFont(QFont("ElysiaOSNew", 14))
            label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
            label.setGeometry(80, 230, 440, 40)
            label.setWordWrap(True)
            label.setVisible(True)
            self.labels.append(label)
            return

        keys_to_display = [
            "state", "warning-level", "energy-full", "energy-full-design",
            "energy-rate", "voltage", "charge-cycles", "percentage",
            "capacity", "technology"
        ]

        y = 80
        for key in keys_to_display:
            label = QLabel("", self.battery_box)
            label.setFont(QFont("ElysiaOSNew", 14))
            label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
            label.setGeometry(80, y, 440, 30)
            label.setWordWrap(True)
            label.setVisible(True)
            self.labels.append(label)
            y += 35

        self.update_info()

    def update_info(self):
        if not self.battery_device:
            return

        battery_info = self.parse_battery_info(self.battery_device)
        keys = [
            "state", "warning-level", "energy-full", "energy-full-design",
            "energy-rate", "voltage", "charge-cycles", "percentage",
            "capacity", "technology"
        ]

        for i, key in enumerate(keys):
            value = battery_info.get(key, "N/A")
            self.labels[i].setText(f"{key.capitalize()}: {value}")

    def show(self):
        self.parent.hide_tiles()
        self.battery_box.setVisible(True)
        self.parent.show_back_button()
        self.update_info()
        self.timer.start(1000)

    def hide(self):
        self.timer.stop()
        self.battery_box.setVisible(False)
        self.parent.hide_back_button()
        self.parent.show_tiles()

File: components/Battery.cpp
#include "Battery.h"
#include "../translations.h"
#include <QDir>
#include <QFont>
#include <QProcess>
#include <QVBoxLayout>
#include <QLabel>
#include <QTimer>

Battery::Battery(QObject *parent) : QObject(parent) {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent);
    if (!parentWidget) return;
    batteryDevice = findBatteryDevice();
    initBatteryUI();
    updateTimer = new QTimer(this);
    updateTimer->setInterval(1000);
    connect(updateTimer, &QTimer::timeout, this, &Battery::updateBatteryStatus);
}

QString Battery::findBatteryDevice() {
    QProcess process;
    process.start("upower", QStringList() << "-e");
    process.waitForFinished();
    QString output = QString::fromUtf8(process.readAllStandardOutput());
    QStringList lines = output.split('\n', Qt::SkipEmptyParts);
    for (const QString &line : lines) {
        if (line.contains("battery_")) return line.trimmed();
    }
    return QString();
}

QMap<QString, QString> Battery::parseBatteryInfo(const QString &device) {
    QMap<QString, QString> info;
    QProcess process;
    process.start("upower", QStringList() << "-i" << device);
    process.waitForFinished();
    QString output = QString::fromUtf8(process.readAllStandardOutput());
    QStringList wantedKeys = {"state", "warning-level", "energy-full", "energy-full-design", "energy-rate", "voltage", "charge-cycles", "percentage", "capacity", "technology"};
    for (const QString &line : output.split('\n')) {
        if (!line.contains(":")) continue;
        QStringList parts = line.trimmed().split(":", Qt::SkipEmptyParts);
        if (parts.size() < 2) continue;
        QString key = parts[0].trimmed();
        QString value = parts[1].trimmed();
        if (wantedKeys.contains(key)) info[key] = value;
    }
    return info;
}

void Battery::initBatteryUI() {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (!parentWidget) return;
    batteryBox = new QWidget(parentWidget);
    batteryBox->setGeometry(500, 100, 600, 500);
    batteryBox->setStyleSheet(
        "background-color: rgba(255, 255, 255, 0.05);"
        "border: 2px solid rgba(255, 255, 255, 0.15);"
        "border-radius: 30px;"
    );
    batteryBox->setVisible(false);
    QVBoxLayout *layout = new QVBoxLayout(batteryBox);
    layout->setContentsMargins(40, 40, 40, 40);
    layout->setSpacing(18);
    QLabel *title = new QLabel(Translations::tr("BATTERY_TITLE"), batteryBox);
    title->setFont(QFont("ElysiaOSNew", 17, QFont::Bold));
    title->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
    title->setAlignment(Qt::AlignCenter);
    layout->addWidget(title);
    if (batteryDevice.isEmpty()) {
        QLabel *label = new QLabel(Translations::tr("BATTERY_TITLE"), batteryBox);
        label->setFont(QFont("ElysiaOSNew", 12));
        label->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
        label->setWordWrap(true);
        label->setAlignment(Qt::AlignCenter);
        layout->addWidget(label);
        infoLabels.clear();
        infoLabels.append(label);
        return;
    }
    QStringList keys = {"state", "warning-level", "energy-full", "energy-full-design", "energy-rate", "voltage", "charge-cycles", "percentage", "capacity", "technology"};
    infoLabels.clear();
    for (const QString &key : keys) {
        QLabel *label = new QLabel("", batteryBox);
        label->setFont(QFont("ElysiaOSNew", 12));
        label->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
        label->setWordWrap(true);
        layout->addWidget(label);
        infoLabels.append(label);
    }
    updateBatteryStatus();
}

void Battery::show() {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (!parentWidget) return;
    if (parentWidget->metaObject()->indexOfMethod("hideTiles()") != -1)
        QMetaObject::invokeMethod(parentWidget, "hideTiles");
    if (parentWidget->metaObject()->indexOfMethod("showBackButton()") != -1)
        QMetaObject::invokeMethod(parentWidget, "showBackButton");
    batteryBox->setVisible(true);
    updateBatteryStatus();
    updateTimer->start();
}

void Battery::hide() {
    batteryBox->setVisible(false);
    updateTimer->stop();
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (!parentWidget) return;
    if (parentWidget->metaObject()->indexOfMethod("hideBackButton()") != -1)
        QMetaObject::invokeMethod(parentWidget, "hideBackButton");
    if (parentWidget->metaObject()->indexOfMethod("showTiles()") != -1)
        QMetaObject::invokeMethod(parentWidget, "showTiles");
}

void Battery::updateBatteryStatus() {
    if (batteryDevice.isEmpty()) return;
    QMap<QString, QString> info = parseBatteryInfo(batteryDevice);
    QStringList keys = {"state", "warning-level", "energy-full", "energy-full-design", "energy-rate", "voltage", "charge-cycles", "percentage", "capacity", "technology"};
    for (int i = 0; i < keys.size() && i < infoLabels.size(); ++i) {
        QString value = info.value(keys[i], "N/A");
        infoLabels[i]->setText(QString("%1: %2").arg(keys[i].replace("-", " ").toUpper(), value));
    }
} 